trust-quorum/types/versions/src/initial/crypto.rs

```diff
@@ -0,0 +1,161 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Cryptographic types for trust quorum.
+
+use gfss::shamir::CombineError;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_with::{hex::Hex, serde_as};
+use slog_error_chain::SlogInlineError;
+
+/// A SHA3-256 digest (32 bytes).
+#[serde_as]
+#[derive(
+    Default,
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+)]
+#[schemars(transparent)]
+pub struct Sha3_256Digest(
+    #[serde_as(as = "Hex")]
+    #[schemars(with = "String")]
+    pub [u8; 32],
+);
+
+impl std::fmt::Debug for Sha3_256Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "sha3 digest: ")?;
+        for v in self.0.as_slice() {
+            write!(f, "{:x?}", v)?;
+        }
+        Ok(())
+    }
+}
+
+/// Some public randomness for cryptographic operations.
+#[serde_as]
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Eq,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+)]
+#[schemars(transparent)]
+pub struct Salt(
+    #[serde_as(as = "Hex")]
+    #[schemars(with = "String")]
+    pub [u8; 32],
+);
+
+/// All possibly relevant __encrypted__ rack secrets for _prior_ committed
+/// configurations.
+#[serde_as]
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+)]
+pub struct EncryptedRackSecrets {
+    /// A random value used to derive the key to encrypt the rack secrets for
+    /// prior committed epochs.
+    pub salt: Salt,
+    #[serde_as(as = "Hex")]
```

Comment:
We seem to be doing this unnecessarily inside the protocol over sprockets between nodes. It only needs to be done at the HTTP level, as we encode types in CBOR over sprockets. This is probably fine, but I"m pointing it out for posterity.

Reply:
I'm not sure we can with derive macros. In the [DB code](https://github.com/oxidecomputer/omicron/blob/main/nexus/db-queries/src/db/datastore/trust_quorum.rs#L679) I manually did the encoding, but that doesn't make much sense here.